### PR TITLE
LUCENE-8746: Refactor EdgeTree 

### PR DESCRIPTION
**Pull request by iverase**
Another try in refactoring edge tree. This PR splits Edge Tree class into two and adds a new interface:

* Component2D: Interface defining an object that knows its bounding box and can perform some spatial operations.
* ComponentTree: An interval tree containing the different components (e.g polygon or line) 
* EdgeTree: An interval tree containing the edges of a components (polygon or line edges)

Unfortunately the PR touches quite a lots of files but most of them are test files. Running benchmark for points results look good, points shows same performance and there is an increase of performance for shapes (we are not computing the bounding box of the triangle many times).
 
 |Approach|Shape|M hits/sec Dev|M hits/sec Base|Diff| QPS  Dev| QPS Base|Diff|Hit count Dev|Hit count Base|Diff|
|--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | 
|points|polyRussia|13.97|13.98|-0%|3.98|3.99|-0%|3508846|3508846| 0%|
|points|poly 10|73.35|71.84| 2%|46.39|45.43| 2%|355809475|355809475| 0%|
|points|polyMedium|8.73|8.67| 1%|106.99|106.24| 1%|2693559|2693559| 0%|
|shapes|polyRussia|6.88|5.73|20%|1.96|1.63|20%|3508846|3508846| 0%|
|shapes|poly 10|27.98|27.03| 4%|17.69|17.10| 4%|355809475|355809475| 0%|
|shapes|polyMedium|2.79|2.75| 2%|34.19|33.66| 2%|2693559|2693559| 0%|

### CODE REVIEW

**Comment by nknize** [comment]
I checked out the PR and got the following test failures:

```
NOTE: reproduce with: ant test  -Dtestcase=TestLatLonPointQueries -Dtests.method=testAllLatEqual -Dtests.seed=DF1A4B86CC81C3AB -Dtests.slow=true -Dtests.badapples=true -Dtests.locale=seh-MZ -Dtests.timezone=Antarctica/Syowa -Dtests.asserts=true -Dtests.file.encoding=UTF-8

NOTE: reproduce with: ant test  -Dtestcase=TestLatLonPointQueries -Dtests.method=testSmallSetPolyWholeMap -Dtests.seed=DF1A4B86CC81C3AB -Dtests.slow=true -Dtests.badapples=true -Dtests.locale=seh-MZ -Dtests.timezone=Antarctica/Syowa -Dtests.asserts=true -Dtests.file.encoding=UTF-8

NOTE: reproduce with: ant test  -Dtestcase=TestLatLonPointQueries -Dtests.method=testLowCardinality -Dtests.seed=DF1A4B86CC81C3AB -Dtests.slow=true -Dtests.badapples=true -Dtests.locale=seh-MZ -Dtests.timezone=Antarctica/Syowa -Dtests.asserts=true -Dtests.file.encoding=UTF-8

NOTE: reproduce with: ant test  -Dtestcase=TestLatLonPointQueries -Dtests.method=testAllLonEqual -Dtests.seed=DF1A4B86CC81C3AB -Dtests.slow=true -Dtests.badapples=true -Dtests.locale=seh-MZ -Dtests.timezone=Antarctica/Syowa -Dtests.asserts=true -Dtests.file.encoding=UTF-8

NOTE: reproduce with: ant test  -Dtestcase=TestLatLonPointQueries -Dtests.method=testRandomMedium -Dtests.seed=DF1A4B86CC81C3AB -Dtests.slow=true -Dtests.badapples=true -Dtests.locale=seh-MZ -Dtests.timezone=Antarctica/Syowa -Dtests.asserts=true -Dtests.file.encoding=UTF-8

NOTE: reproduce with: ant test  -Dtestcase=TestLatLonPointQueries -Dtests.method=testRandomTiny -Dtests.seed=DF1A4B86CC81C3AB -Dtests.slow=true -Dtests.badapples=true -Dtests.locale=seh-MZ -Dtests.timezone=Antarctica/Syowa -Dtests.asserts=true -Dtests.file.encoding=UTF-8
```

**Comment by iverase** [comment]
@nknize totally right, some bug I introduce when trying to improve it. Now it should be fixed.

File: lucene/core/src/java/org/apache/lucene/geo/EdgeTree.java [hunk 1]
```diff
@@ -269,138 +157,109 @@ boolean crossesTriangle(double ax, double ay, double bx, double by, double cx, d
...
    */
-  private static Edge createTree(double[] lats, double[] lons) {
-    Edge edges[] = new Edge[lats.length - 1];
-    for (int i = 1; i < lats.length; i++) {
-      double lat1 = lats[i-1];
-      double lon1 = lons[i-1];
-      double lat2 = lats[i];
-      double lon2 = lons[i];
-      edges[i - 1] = new Edge(lat1, lon1, lat2, lon2, Math.min(lat1, lat2), Math.max(lat1, lat2));
+  public static EdgeTree createTree(double[] x, double[] y) {
```
Comment (nknize) [inline lucene/core/src/java/org/apache/lucene/geo/EdgeTree.java]:
```suggestion
  protected static EdgeTree createTree(double[] x, double[] y) {
```

File: lucene/core/src/java/org/apache/lucene/geo/EdgeTree.java [hunk 2]
```diff
@@ -269,138 +157,109 @@ boolean crossesTriangle(double ax, double ay, double bx, double by, double cx, d
...
-    //check the bounding box because if the triangle is degenerated, e.g points and lines, we need to filter out
-    //coplanar points that are not part of the triangle.
-    if (x >= minX && x <= maxX && y >= minY && y <= maxY ) {
-      int a = orient(x, y, ax, ay, bx, by);
-      int b = orient(x, y, bx, by, cx, cy);
-      if (a == 0 || b == 0 || a < 0 == b < 0) {
-        int c = orient(x, y, cx, cy, ax, ay);
-        return c == 0 || (c < 0 == (b < 0 || a < 0));
+  /** Returns true if the line crosses any edge in this edge subtree */
+  public boolean crossesLine(double minX, double maxX, double minY, double maxY, double a2x, double a2y, double b2x, double b2y) {
```
Comment (nknize) [inline lucene/core/src/java/org/apache/lucene/geo/EdgeTree.java]:
```suggestion
  protected boolean crossesLine(double minX, double maxX, double minY, double maxY, double a2x, double a2y, double b2x, double b2y) {
```

File: lucene/core/src/java/org/apache/lucene/geo/EdgeTree.java [hunk 3]
```diff
@@ -34,232 +31,123 @@
...
-        double ey = lat2;
-        double dx = lon1;
-        double ex = lon2;
-
-        // compute remaining bounding box of triangle
-        double triMinLon = StrictMath.min(StrictMath.min(ax, bx), cx);
-        double triMaxLat = StrictMath.max(StrictMath.max(ay, by), cy);
-        double triMaxLon = StrictMath.max(StrictMath.max(ax, bx), cx);
+  /** Returns true if the triangle crosses any edge in this edge subtree */
+  public boolean crossesTriangle(double minX, double maxX, double minY, double maxY,
```
Comment (nknize) [inline lucene/core/src/java/org/apache/lucene/geo/EdgeTree.java]:
```suggestion
  protected boolean crossesTriangle(double minX, double maxX, double minY, double maxY,
```

File: lucene/core/src/java/org/apache/lucene/geo/EdgeTree.java [hunk 4]
```diff
@@ -34,232 +31,123 @@
...
-      } else {
-        comparator = (left, right) -> {
-          int ret = Double.compare(left.minLat, right.minLat);
-          if (ret == 0) {
-            ret = Double.compare(left.maxY, right.maxY);
-          }
-          return ret;
-        };
+  /** returns true if the provided x, y point lies on the line */
+  public boolean isPointOnLine(double x, double y) {
```
Comment (nknize) [inline lucene/core/src/java/org/apache/lucene/geo/EdgeTree.java]:
```suggestion
  protected boolean isPointOnLine(double x, double y) {
```

File: lucene/core/src/java/org/apache/lucene/geo/Component2D.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  /** Compute whether the bounding boxes are disjoint **/
+  static  boolean disjoint(double minX1, double maxX1, double minY1, double maxY1, double minX2, double maxX2, double minY2, double maxY2) {
+    return (maxY1 < minY2 || minY1 > maxY2 || maxX1 < minX2 || minX1 > maxX2);
+  }
+
+  /** Compute whether the first bounding box 1 is within the second bounding box **/
+  static boolean within(double minX1, double maxX1, double minY1, double maxY1, double minX2, double maxX2, double minY2, double maxY2) {
+    return (minY2 < minY1 && maxY2 > maxY1&& minX2 < minX1 && maxX2 > maxX1);
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/core/src/java/org/apache/lucene/geo/Component2D.java]: good call! indeed this is a performance bug.

File: lucene/core/src/java/org/apache/lucene/document/LatLonPointInPolygonQuery.java [hunk 6]
```diff
@@ -133,28 +133,25 @@ public Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
...
-            return tree.relate(cellMinLat, cellMaxLat, cellMinLon, cellMaxLon);
+            return tree.relate(cellMinLon, cellMaxLon, cellMinLat, cellMaxLat);
           }
         };
   }
 
   @Override
   public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
 
+    final Component2D tree = Polygon2D.create(polygons);
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/core/src/java/org/apache/lucene/document/LatLonPointInPolygonQuery.java]: Not sure if we should do that. I think we create one ScorerSupplier per segment so moving that logic to the IntersectVisitor means that we will creating this objects per segment.

**Comment by iverase** [comment]
@nkine do you think we can move this forward?

File: lucene/core/src/java/org/apache/lucene/document/LatLonPointInPolygonQuery.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (nknize) [reply lucene/core/src/java/org/apache/lucene/document/LatLonPointInPolygonQuery.java]: :+1:  Thanks for reminding me. I ran into that same problem ages ago as well.

**Pull request merged**